Request workshop rework for FAC17 (#8)
foundersandcoders/mc-request-module-workshop: README.md

> # Morning challenge - Build the Request module

Reviewer: ```suggestion
# Morning challenge - Build a request module
```

> This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and methods. Now that you are in the backend we would like you to learn how to make similar request using the Node core [http module](https://nodejs.org/api/http.html). This workshop is intended to familiarise you with how backend requests are made, and practice using callbacks.

Reviewer: ```suggestion
This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and methods. Now that you are in the backend we would like you to learn how to make similar requests using the Node core [http module](https://nodejs.org/api/http.html). This workshop is intended to familiarise you with how backend requests are made, and practice using callbacks.
```

> ### Time Required: 1 hour

Reviewer: Not sure this section is needed. Students should know how a morning challenge goes by Week 5

Author: [Removed](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/922e75e8d598eea525c7484ac5217f0563fa7ca9)

  - 10 mins to read challenge instructions
  - 40 mins to complete the challenge in pairs
  - 10 mins to regroup and talk over the solution

### Learning outcome
To become familiar with making API calls from the backend using the Node core [http module](https://nodejs.org/api/http.html).

> ### When would I need to make an API call from the backend?

Reviewer: Maybe the examples should highlight why students might want to make an API call _from the backend_. E.g bypassing CORS or hiding API keys

Author: [Updated](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/d3c8b3c6b7bc888a43eb793c02c47529fea5d22e)

Example 1 - I would like to make a HTTP request to the [GitHub API](https://developer.github.com/v3/) to get my profile photo URL.

Example 2 - I would like to make a HTTP request to the [TFL API](https://api.tfl.gov.uk/), to find out which buses go between London Bridge and Finsbury Park.

### The challenge

Create a function called `myRequest` which makes a request to an API url using an http GET request.

It should take two arguments: the url to make the request to, and a callback function.

> The callback function will have two arguments:

Reviewer: ```suggestion
The callback function should take two arguments:
```

>   1. error (String: if an error occurred)

Reviewer: ```suggestion
  1. error (null if no error occurred; an error object if an error occurred)
```

>   2. response (Object: includes the body & statusCode of the request)

Reviewer 2: It would be helpful to add a hint that you should be constructing your own response object here

Reviewer: They don't _have_ to make a new object—the Node `response` object already has the `statusCode` so they could just add the `body` to it and use that

Author: Probably deserves some kind of note/hint, it's something we both got stuck on trying out the new workshop

Reviewer: I guess this is why the old workshop had the callback take 3 arguments: `(error, response, body)`

A note/hint is probably a good idea

Author: I've added a [hint](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/1774c5c438a3338130fcd2efa4634071aee071e1)

**Hint**: relate to your knowledge of streams and sending data in chunks.

To complete this challenge you will need to use the [`http.get`](https://nodejs.org/api/http.html#http_http_get_options_callback) method of the node core `http` module.

### Start here!
1) Clone this repo & run `npm install`

> 2) Run `npm test` in your command line to start the test watcher. The tests should fail as your function doesn't exist yet

Reviewer: ```suggestion
2) Run `npm test` in your command line to start the test watcher. The tests should fail as your function doesn't do anything yet
```

3) Open `app.js` and write your function!

> ### Tests

Reviewer: Not sure if this section will be helpful or just confusing. They can always read the tests if the instructions above aren't clear enough

Author: When @bantek89 tried this out for practice I think she found the tests useful? We could see how FAC17 find it and remove if it is confusing?

Reviewer: I agree that the tests are useful, I just don't know if a section explaining what the tests do adds much. If the tests themselves are confusing then maybe they should have comments next to them in the code?

Author: OK I've removed that section of the readme

There are two test sets in `app.test.js` (the second is commented out). The first test will run three assertions: 
1) Checks there are no errors
2) Checks that the API response passed to the callback has a status code of 200
3) Checks that the API response passed to the callback has the correct json

The second, commented-out test will check your function works with https requests (bonus round, see below...). Uncomment it when you're ready to test in the bonus round!

### Bonus round

> You can alter your myRequest function to check which protocol the url is using (http or https) and make the suitable request. You will need to require in the `https` Node core module. You can test this by uncommenting the second test in `app.test.js`.

Reviewer: Maybe add an intro to this asking them to try making a request to the Pokeapi (or anything that uses https) so they can see it fail

Author: I've added a more detailed intro, how's [this](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/861b2debb769660884c075ecf50a3078c1daa126)?

> You can alter your myRequest function to check which protocol the url is using (http or https) and make the suitable request. You will need to require in the `https` Node core module. You can test this by uncommenting the second test in `app.test.js`.

Reviewer: ```suggestion
You can alter your `myRequest` function to check which protocol the url is using (http or https) and use the appropriate Node core module. You will need to require in `https`. You can test this by uncommenting the second test in `app.test.js`.
```
